portia_server/portia_api/utils/extract.py

```diff
@@ -0,0 +1,109 @@
+import json
+import logging
+
+from twisted.internet import defer
+from twisted.web.client import getPage
+
+from scrapy import Request
+from scrapy.http import HtmlResponse
+from scrapy.settings import Settings
+
+from slybot.spider import IblSpider
+
+from crochet import setup, wait_for, TimeoutError
+setup()
+
+
+class FetchError(Exception):
+    status = 400
+
+    def __init__(self, errors):
+        for error in errors:
+            print(vars(error))
+        self.errors = errors
+        self.message = str(self)
+
+    def __str__(self):
+        return '\n'.join(e.getErrorMessage() for e in self.errors)
+
+
+def get_page(times, url):
+    errors = []
+    deferred = defer.Deferred()
+
+    def run():
+        d = getPage(url)
+        d.addCallbacks(lambda html: deferred.callback((url, html)), error)
+
+    def error(error):
+        errors.append(error)
+        retry = True if len(errors) < times else False
+        logging.warn('Failed to get %s %d times, %s', url, len(errors),
+                     'retrying' if retry else 'stop')
+        run() if retry else deferred.errback((url, errors))
+    run()
+    return deferred
+
+
+def _load_urls(urls):
+    deferreds = []
+    for url in urls:
+        deferreds.append(get_page(3, url.encode('utf-8')))
+    return defer.DeferredList(deferreds)
+
+
+@wait_for(timeout=50)
+def load_urls(urls):
+    return _load_urls(urls)
+
+
+class Pages(object):
+    def __init__(self, urls, user=None):
```

Comment:
It could be useful for custom implementations but I'll remove it and add a separate build_pages method